File: 2015/day1.js
const fs = require("fs");

const part1 = input => input
      .split('')
      .reduce((acc, bracket) => {
        return acc + (bracket === '(' ? 1 : -1)
      }, 0)

const part2 = input => {
  let floor = 0
  const inputArr = input.split('')
  for (let idx = 0; idx < inputArr.length; idx++) {
    floor += (inputArr[idx] === '(' ? 1 : -1)
    if (floor === -1) {
      return idx + 1
    }
  }
}

const input = fs.readFileSync("inputs/day1.txt", { encoding: "utf-8" }).trim()
console.log(`Part 1: ${part1(input)}`)
console.log(`Part 2: ${part2(input)}`)

File: 2019/day1.exs
defmodule Day1 do
  @input "inputs/1"

  defp normalize(content) do
    content
    |> String.split("\n", trim: true)
    |> Stream.map(&(String.to_integer(&1)))
  end

  defp get_input do
    {:ok, content} = File.read(@input)
    normalize(content)
  end

  defp fuel_required(mass) do
    Integer.floor_div(mass, 3) - 2
  end

  def part1() do
    get_input()
    |> Stream.map(&fuel_required/1)
    |> Enum.sum
  end

  defp total_required(mass) do
    Stream.iterate(mass, &fuel_required/1)
    |> Stream.take_while(&(&1 > 0))
    |> Stream.drop(1)
    |> Enum.sum
  end

  def part2() do
    get_input()
    |> Stream.map(&total_required/1)
    |> Enum.sum
  end
end

IO.puts("Part 1: #{Day1.part1()}")
IO.puts("Part 2: #{Day1.part2()}")
